modules/sdk/src/main/scala/org/tessellation/sdk/domain/healthcheck/consensus/HealthCheckConsensus.scala

```diff
@@ -204,17 +209,20 @@ abstract class HealthCheckConsensus[
 
              def participate = participateInRound(proposal.key, proposal.roundId)
 
-             inProgressRound
-               .map(_.processProposal(proposal))
-               .orElse {
-                 historicalRound.map(handleProposalForHistoricalRound(proposal))
-               }
-               .getOrElse {
-                 if (depth > 0)
-                   participate.flatMap(_ => handleProposal(proposal, depth - 1))
-                 else
-                   (new Throwable("Unexpected recursion!")).raiseError[F, Unit]
-               }
+             inProgressRound.map { r =>
+               r.hasProposal(proposal.owner)
+                 .ifM(
+                   waitingProposals.update(_ + proposal),
+                   r.processProposal(proposal)
+                 )
+             }.orElse {
+               historicalRound.map(handleProposalForHistoricalRound(proposal))
```

Comment:
Shouldn't we first check the historical rounds and only if the proposal is not matching the historical rounds, then try process it for the in progress ones?

Reply:
Good point :+1: Changed